nifi-nar-bundles/nifi-extension-utils/nifi-processor-utils/src/main/java/org/apache/nifi/processor/util/list/ListedEntityTracker.java

```diff
@@ -0,0 +1,322 @@
+            "Regardless of entities timestamp, all existing entities will be listed at the initial listing activity.");
+    private static final AllowableValue INITIAL_LISTING_TARGET_WINDOW = new AllowableValue("window", "Tracking Time Window",
+            "Ignore entities having timestamp older than the specified 'Tracking Time Window' at the initial listing activity.");
+
+    public static final PropertyDescriptor INITIAL_LISTING_TARGET = new PropertyDescriptor.Builder()
+            .name("et-initial-listing-target")
+            .displayName("Entity Tracking Initial Listing Target")
+            .description(format("Specify how initial listing should be handled." +
+                    " %s", NOTE))
+            .allowableValues(INITIAL_LISTING_TARGET_WINDOW, INITIAL_LISTING_TARGET_ALL)
+            .defaultValue(INITIAL_LISTING_TARGET_WINDOW.getValue())
+            .build();
+
+    public static final PropertyDescriptor NODE_IDENTIFIER = new PropertyDescriptor.Builder()
+            .name("et-node-identifier")
+            .displayName("Entity Tracking Node Identifier")
+            .description(format("The configured value will be appended to the cache key" +
+                    " so that listing state can be tracked per NiFi node rather than cluster wide" +
+                    " when tracking state is scoped to LOCAL. %s", NOTE))
+            .addValidator(StandardValidators.NON_EMPTY_VALIDATOR)
+            .expressionLanguageSupported(ExpressionLanguageScope.VARIABLE_REGISTRY)
+            .defaultValue("${hostname()}")
+            .build();
+
+    private static Supplier<Long> CURRENT_TIMESTAMP = System::currentTimeMillis;
+
+    private Serializer<String> stringSerializer = (v, o) -> o.write(v.getBytes(StandardCharsets.UTF_8));
+    private Serializer<Map<String, ListedEntity>> listedEntitiesSerializer
+            = (v, o) -> objectMapper.writeValue(new GZIPOutputStream(o), v);
+    private Deserializer<Map<String, ListedEntity>> listedEntitiesDeserializer
+            = v -> (v == null || v.length == 0) ? null
+            : objectMapper.readValue(new GZIPInputStream(new ByteArrayInputStream(v)), new TypeReference<Map<String, ListedEntity>>() {});
+
+    private final String componentId;
+    private final ComponentLog logger;
+    private final Scope scope;
+
+    /*
+     * The nodeId and mapCacheClient being used at the previous trackEntities method execution is captured,
+     * so that it can be used when resetListedEntities is called.
+     */
+    private String nodeId;
+    private DistributedMapCacheClient mapCacheClient;
+
+    ListedEntityTracker(String componentId, ComponentLog logger, Scope scope) {
+        this.componentId = componentId;
+        this.logger = logger;
+        this.scope = scope;
+    }
+
+    static void validateProperties(ValidationContext context, Collection<ValidationResult> results, Scope scope) {
+        final Consumer<PropertyDescriptor> validateRequiredProperty = property -> {
```

Comment:
I just wanted to avoid adding another private method just for this validateProperties method. But I agree with you, private method would increase readability. Updated.